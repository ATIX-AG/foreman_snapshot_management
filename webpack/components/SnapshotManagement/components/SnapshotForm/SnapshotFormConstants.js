export const GET_SNAPSHOT_CREATE_REQUEST = 'GET_SNAPSHOT_CREATE_REQUEST';
export const GET_SNAPSHOT_CREATE_SUCCESS = 'GET_SNAPSHOT_CREATE_SUCCESS';
export const GET_SNAPSHOT_CREATE_FAILURE = 'GET_SNAPSHOT_CREATE_FAILURE';

export const SNAPSHOT_CREATE_URL = '/api/hosts/:hostId/snapshots';
