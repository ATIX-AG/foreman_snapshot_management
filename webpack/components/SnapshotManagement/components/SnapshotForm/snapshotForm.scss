textarea.form-control {
  resize: vertical;
}
