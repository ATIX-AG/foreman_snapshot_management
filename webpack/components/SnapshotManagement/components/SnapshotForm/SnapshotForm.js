import React from 'react';
import PropTypes from 'prop-types';
import * as Yup from 'yup';

import { translate as __ } from 'foremanReact/common/I18n';
// import CheckBox from 'foremanReact/components/common/forms/Checkbox';
// import TextInput from 'foremanReact/components/common/forms/TextInput';
import ForemanForm from 'foremanReact/components/common/forms/ForemanForm';
import TextField from 'foremanReact/components/common/forms/TextField';

import { SNAPSHOT_CREATE_URL } from './SnapshotFormConstants';
import './snapshotForm.scss';

const SnapshotForm = ({
  hostId,
  initialValues,
  submitForm,
  capabilities,
  ...props
}) => {
  let nameValidation = Yup.string().max(80, 'Too Long!');
  if (capabilities.limitSnapshotNameFormat)
    nameValidation = nameValidation
      .min(2, 'Too Short!')
      .matches(
        /^[A-z]\w+$/,
        __(
          'Name must contain at least 2 characters starting with alphabet. Valid characters are A-Z a-z 0-9 _'
        )
      )
      .max(40, 'Too Long!');

  const validationSchema = Yup.object().shape({
    name: nameValidation.required('is required'),
    description: Yup.string(),
    includeRam: Yup.bool(),
  });

  const handleSubmit = async (values, actions) => {
    const submitValues = {
      include_ram: values.includeRam || false,
      snapshot: {
        name: values.name,
        description: values.description || '',
      },
    };

    await submitForm({
      item: 'Snapshot',
      url: SNAPSHOT_CREATE_URL.replace(':hostId', hostId),
      values: submitValues,
      message: __('Snapshot successfully created!'),
    });
    actions.setSubmitting(false);
    props.setModalClosed();
  };

  return (
    <ForemanForm
      onSubmit={(values, actions) => handleSubmit(values, actions)}
      initialValues={initialValues}
      validationSchema={validationSchema}
      onCancel={() => props.setModalClosed()}
    >
      <TextField
        name="name"
        type="text"
        required="true"
        label={__('Name')}
        inputClassName="col-md-8"
      />
      <TextField
        name="description"
        type="textarea"
        label={__('Description')}
        inputClassName="col-md-8"
      />
      <TextField
        type="checkbox"
        name="includeRam"
        label={__('Include RAM')}
        inputClassName="col-md-8"
      />
    </ForemanForm>
  );
};

SnapshotForm.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  hostId: PropTypes.number.isRequired,
  initialValues: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    includeRam: PropTypes.bool.isRequired,
  }),
  submitForm: PropTypes.func.isRequired,
  setModalClosed: PropTypes.func,
  capabilities: PropTypes.shape({
    limitSnapshotNameFormat: PropTypes.bool,
  }),
};

SnapshotForm.defaultProps = {
  className: '',
  children: null,
  initialValues: {
    name: '',
    description: '',
    includeRam: false,
  },
  setModalClosed: () => {},
  capabilities: {
    limitSnapshotNameFormat: false,
  },
};

export default SnapshotForm;
