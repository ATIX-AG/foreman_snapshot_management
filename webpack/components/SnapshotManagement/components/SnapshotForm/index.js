import React from 'react';
import { useDispatch } from 'react-redux';

import { submitForm } from 'foremanReact/redux/actions/common/forms';

import SnapshotForm from './SnapshotForm';

const WrappedSnapshotForm = props => {
  const dispatch = useDispatch();

  return (
    <SnapshotForm
      submitForm={(...args) => dispatch(submitForm(...args))}
      {...props}
    />
  );
};

export default WrappedSnapshotForm;
