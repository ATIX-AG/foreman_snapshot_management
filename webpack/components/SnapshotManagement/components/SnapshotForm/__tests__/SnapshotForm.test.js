import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';

import SnapshotForm from '../SnapshotForm';

const fixtures = {
  'render without optional Props': { hostId: 42, submitForm: () => null },
  'render with optional Props': {
    hostId: 42,
    submitForm: () => null,
    initialValues: {
      name: 'Snapshot1',
      description: 'Hello World',
      includeRam: true,
    },
  },
};

describe('SnapshotForm', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(SnapshotForm, fixtures));
});
