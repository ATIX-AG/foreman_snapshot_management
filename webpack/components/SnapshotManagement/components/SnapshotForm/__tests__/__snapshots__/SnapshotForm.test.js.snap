// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnapshotForm rendering render 1`] = `
<ForemanForm
  enableReinitialize={false}
  initialValues={
    Object {
      "description": "",
      "includeRam": false,
      "name": "",
    }
  }
  onCancel={[Function]}
  onSubmit={[Function]}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "includeRam",
        "description",
        "name",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "fields": Object {
        "description": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "includeRam": BooleanSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "boolean",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "boolean",
        },
        "name": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
      "type": "object",
    }
  }
>
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Name"
    name="name"
    required="true"
    type="text"
  />
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Description"
    name="description"
    required={false}
    type="textarea"
  />
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Include RAM"
    name="includeRam"
    required={false}
    type="checkbox"
  />
</ForemanForm>
`;

exports[`SnapshotForm rendering render with limitSnapshotNameFormat capability 1`] = `
<ForemanForm
  enableReinitialize={false}
  initialValues={
    Object {
      "description": "",
      "includeRam": false,
      "name": "",
    }
  }
  onCancel={[Function]}
  onSubmit={[Function]}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "includeRam",
        "description",
        "name",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "fields": Object {
        "description": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "includeRam": BooleanSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "boolean",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "boolean",
        },
        "name": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "matches": false,
            "max": true,
            "min": true,
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
      "type": "object",
    }
  }
>
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Name"
    name="name"
    required="true"
    type="text"
  />
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Description"
    name="description"
    required={false}
    type="textarea"
  />
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Include RAM"
    name="includeRam"
    required={false}
    type="checkbox"
  />
</ForemanForm>
`;

exports[`SnapshotForm rendering render with optional Props 1`] = `
<ForemanForm
  enableReinitialize={false}
  initialValues={
    Object {
      "description": "Hello World",
      "includeRam": true,
      "name": "Snapshot1",
    }
  }
  onCancel={[Function]}
  onSubmit={[Function]}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "includeRam",
        "description",
        "name",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Array [],
        "refs": Array [],
      },
      "fields": Object {
        "description": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "includeRam": BooleanSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {},
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "boolean",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "boolean",
        },
        "name": StringSchema {
          "_blacklist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Array [],
            "refs": Array [],
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
      "type": "object",
    }
  }
>
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Name"
    name="name"
    required="true"
    type="text"
  />
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Description"
    name="description"
    required={false}
    type="textarea"
  />
  <TextField
    className=""
    inputClassName="col-md-8"
    label="Include RAM"
    name="includeRam"
    required={false}
    type="checkbox"
  />
</ForemanForm>
`;
